### PR TITLE
import: tune SstFileWriter options to reduce CPU usage

### PR DESCRIPTION
**Pull request by lonng**
Signed-off-by: Lonng <chris@lonng.org>

## What have you changed? (mandatory)

The `zstd` is a CPU-tense compaction algorithm, which makes other workers have less chance to do some works. CPU will become the bottleneck of TiKV-Importer.

## What are the type of the changes? (mandatory)

Change the compaction algorithm to `lz4` to make a balance between CPU and bandwidth.

- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

11T test case, index engine file import throughput increase 250MB/s to 1GB/s in our 10G network card.

### CODE REVIEW

File: src/import/engine.rs [hunk 1]
```diff
@@ -36,6 +36,8 @@ use crate::import::stream::SSTFile;
 use engine::rocks::util::security::encrypted_env_from_cipher_file;
 use tikv_util::security::SecurityConfig;
 
+const DISABLED: i32 = i32::MAX;
```
Comment (kennytm) [inline src/import/engine.rs]: Why this needs to be moved out ?

File: src/import/engine.rs [hunk 2]
```diff
@@ -254,6 +256,13 @@ impl SSTWriter {
         })
     }
 
+    fn tune_config(env: Arc<Env>, cf_opts: &mut ColumnFamilyOptions) {
+        use DBCompressionType::Lz4;
+        cf_opts.set_env(env);
+        cf_opts.compression_per_level(&[Lz4, Lz4, Lz4, Lz4, Lz4, Lz4, Lz4]);
+        cf_opts.compaction_priority(CompactionPriority::OldestLargestSeqFirst);
```
Comment (zhangjinpeng1987) [inline src/import/engine.rs]: What consideration makes you use this priority?

File: src/import/engine.rs [hunk 3]
```diff
@@ -254,6 +256,13 @@ impl SSTWriter {
         })
     }
 
+    fn tune_config(env: Arc<Env>, cf_opts: &mut ColumnFamilyOptions) {
+        use DBCompressionType::Lz4;
+        cf_opts.set_env(env);
+        cf_opts.compression_per_level(&[Lz4, Lz4, Lz4, Lz4, Lz4, Lz4, Lz4]);
```
Comment (zhangjinpeng1987) [inline src/import/engine.rs]: IMO, compression can increase CPU usage.

**Comment by zhangjinpeng1987** [comment]
@lonng force push is not recommend.

**Comment by lonng** [comment]
> @lonng force push is not recommend.

Got it.

File: src/import/engine.rs (hunk 2 above)
Comment (lonng) [reply src/import/engine.rs]: Sorry, forgot to delete this line.

File: src/import/engine.rs (hunk 1 above)
Comment (lonng) [reply src/import/engine.rs]: Reverted

File: src/import/engine.rs (hunk 3 above)
Comment (kennytm) [reply src/import/engine.rs]: The SstFileWriter only uses the [compression of the last level](https://github.com/facebook/rocksdb/blob/18864567c8a1139d2fc83cda24587043d5beaf84/table/sst_file_writer.cc#L197-L200) (Level 6). The default setting uses `Zstd` which [has higher compression ratio but is slower](https://stackoverflow.com/a/40151577). 

For sure we could set it to `&[No, No, No, No, No, No, Lz4]` but the outcome is the same.

Comment (lonng) [reply src/import/engine.rs]: @kennytm addressed

File: src/import/engine.rs [hunk 4]
```diff
@@ -236,13 +236,14 @@ impl SSTWriter {
...
-        default_opts.set_env(Arc::clone(&env));
+        Self::tune_config(Arc::clone(&env), &mut default_opts);
         let mut default = SstFileWriter::new(EnvOptions::new(), default_opts);
         default.open(&format!("{}{}.{}:default", path, MAIN_SEPARATOR, uuid))?;
 
         // Creates a writer for write CF
         let mut write_opts = db_cfg.writecf.build_opt(&cache);
-        write_opts.set_env(Arc::clone(&env));
+        Self::tune_config(Arc::clone(&env), &mut write_opts);
+
```
Comment (lonng) [inline src/import/engine.rs]:
```suggestion
```

File: src/import/engine.rs [hunk 5]
```diff
@@ -256,6 +256,12 @@ impl SSTWriter {
         })
     }
 
+    fn tune_config(env: Arc<Env>, cf_opts: &mut ColumnFamilyOptions) {
+        use DBCompressionType::{Lz4, No};
+        cf_opts.set_env(env);
+        cf_opts.compression_per_level(&[No, No, No, No, No, No, Lz4]);
```
Comment (siddontang) [inline src/import/engine.rs]: this may cause a higher IO which may cause IO becoming the bottleneck, is it?

Comment (lonng) [reply src/import/engine.rs]: The IO usage will increase but will not be the bottleneck. It's a trade-off.

Comment (zhangjinpeng1987) [reply src/import/engine.rs]: It's better if it is configurable.

Comment (lonng) [reply src/import/engine.rs]: We are decreasing the unnecessary configuration of TiKV-Importer to improve usability. I don't think we really need to config the compaction algorithm.

**Comment by lonng** [comment]
ping @zhangjinpeng1987 @kennytm @Connor1996

File: src/import/engine.rs (hunk 5 above)
Comment (lonng) [reply src/import/engine.rs]: Addressed.
There is an exists configuration to archive this.

**Pull request merged**